src/pages/workspaces/workspace/notebooks/NotebookLauncher.js

```diff
@@ -32,37 +91,219 @@ const NotebookLauncher = _.flow(
     title: _.get('notebookName'),
     showTabBar: false
   })
-)(({ queryParams = {}, notebookName, workspace, workspace: { accessLevel, canCompute }, cluster, refreshClusters, onRequesterPaysError }, ref) => {
-  return (Utils.canWrite(accessLevel) && canCompute && !queryParams['read-only']) ?
-    h(NotebookEditor, { notebookName, workspace, cluster, refreshClusters }) :
-    h(Fragment, [
-      h(ReadOnlyMessage, { notebookName, workspace }),
-      h(NotebookPreviewFrame, { notebookName, workspace, onRequesterPaysError })
+)(
+  ({ queryParams = {}, notebookName, workspace, workspace: { workspace: { namespace, name }, accessLevel, canCompute }, cluster, refreshClusters, onRequesterPaysError }, ref) => {
+    const [createOpen, setCreateOpen] = useState(false)
+    // Status note: undefined means still loading, null means no cluster
+    const clusterStatus = cluster && cluster.status
+    const [busy, setBusy] = useState()
+    const mode = queryParams['mode']
+
+    return h(Fragment, [
+      (Utils.canWrite(accessLevel) && canCompute && !!mode && clusterStatus === 'Running') ?
+        h(NotebookEditorFrame, { key: cluster.clusterName, workspace, cluster, notebookName, mode }) :
+        h(Fragment, [
+          h(PreviewHeader, { queryParams, cluster, refreshClusters, notebookName, workspace, readOnlyAccess: !(Utils.canWrite(accessLevel) && canCompute) }),
+          h(NotebookPreviewFrame, { notebookName, workspace })
+        ]),
+      mode && h(ClusterKicker, { cluster, refreshClusters, onNullCluster: () => { setCreateOpen(true) } }),
+      mode && h(ClusterStatusMonitor, { cluster, onClusterStoppedRunning: () => { chooseMode(undefined) } }),
+      createOpen && h(NewClusterModal, {
+        namespace, currentCluster: cluster,
+        onCancel: () => setCreateOpen(false),
```

Comment:
We probably want to also reset back to preview mode here, otherwise you end up in a state where clicking Edit does nothing, since you're already in Edit mode.